src/ol_infrastructure/applications/keycloak/__main__.py

```diff
@@ -0,0 +1,418 @@
+target_vpc_name = keycloak_config.get("target_vpc") or f"{stack_info.env_prefix}_vpc"
+target_vpc = network_stack.require_output(target_vpc_name)
+target_vpc_id = target_vpc["id"]
+
+mitodl_zone_id = dns_stack.require_output("odl_zone_id")
+
+# TODO MD 20230206
+# This might be needed in the future but right now it just causes errors
+secrets = read_yaml_secrets(Path(f"keycloak/data.{stack_info.env_suffix}.yaml"))
+
+aws_config = AWSBase(
+    tags={"OU": "operations", "Environment": f"operations-{stack_info.env_suffix}"}
+)
+consul_provider = get_consul_provider(stack_info)
+
+env_name = f"{stack_info.env_prefix}-{stack_info.env_suffix}"
+
+# IAM and instance profile
+keycloak_instance_role = iam.Role(
+    f"keycloak-instance-role-{env_name}",
+    assume_role_policy=json.dumps(
+        {
+            "Version": "2012-10-17",
+            "Statement": {
+                "Effect": "Allow",
+                "Action": "sts:AssumeRole",
+                "Principal": {"Service": "ec2.amazonaws.com"},
+            },
+        }
+    ),
+    path="/ol-operations/keycloak/role/",
+    tags=aws_config.tags,
+)
+
+iam.RolePolicyAttachment(
+    f"keycloak-describe-instance-role-policy-{env_name}",
+    policy_arn=policy_stack.require_output("iam_policies")["describe_instances"],
+    role=keycloak_instance_role.name,
+)
+iam.RolePolicyAttachment(
+    "keycloak-route53-records-permission",
+    policy_arn=policy_stack.require_output("iam_policies")["route53_odl_zone_records"],
+    role=keycloak_instance_role.name,
+)
+
+keycloak_server_instance_profile = iam.InstanceProfile(
+    f"keycloak-instance-profile-{env_name}",
+    role=keycloak_instance_role.name,
+    path="/ol-operations/keycloak/profile/",
+)
+
+# Network Access Control
+
+# Create various security groups
+keycloak_server_security_group = ec2.SecurityGroup(
+    f"keycloak-server-security-group-{env_name}",
+    name=f"keycloak-server-{target_vpc_name}-{env_name}",
+    description="Access control for keycloak servers",
+    ingress=[
+        ec2.SecurityGroupIngressArgs(
+            protocol="tcp",
+            from_port=DEFAULT_HTTPS_PORT,
+            to_port=DEFAULT_HTTPS_PORT,
+            cidr_blocks=["0.0.0.0/0"],
+            description=f"Allow traffic to the keycloak server on port {DEFAULT_HTTPS_PORT}",  # noqa: E501
+        ),
+    ],
+    egress=default_egress_args,
+    vpc_id=target_vpc_id,
+    tags=aws_config.tags,
+)
+
+keycloak_database_security_group = ec2.SecurityGroup(
+    f"keycloak-database-security-group-{env_name}",
+    name=f"keycloak-database-{target_vpc_name}-{env_name}",
+    description="Access control for the keycloak database",
+    ingress=[
+        ec2.SecurityGroupIngressArgs(
+            security_groups=[
+                keycloak_server_security_group.id,
+                consul_stack.require_output("security_groups")["consul_server"],
+                vault_stack.require_output("vault_server")["security_group"],
```

Comment:
We'll likely also want to allow access from the data integrator SG for being able to incorporate user info into the data platform.